### PR TITLE
prescribe aad-pod-identity installation in kube-system

### LINKED ISSUE #467: [Azure Firewall + add-pod-identity] watching problem
**Describe the bug**

## Symptom

when restricting egress traffic in an aks cluster using Azure Firewall, aad pod identity starts failing while trying to watch.

### Azure Firewall logs

the following is the only denied HTTPS request:

```
"msg":"HTTPS request from 10.10.0.4:45140. Action: Deny. Reason: SNI TLS extension was missing."
```

from user's voice: https://feedback.azure.com/forums/217313-networking/suggestions/38623357-disable-sni-tls-extension-check-on-azure-firewall

### nmi pod

`kubectl get po nmi-szsqt -o wide`

```
NAME        READY   STATUS    RESTARTS   AGE     IP          NODE                                NOMINATED NODE   READINESS GATES
nmi-szsqt   1/1     Running   0          3h51m   10.10.0.4   aks-agentpool-84518513-vmss000000   <none>           <none>
```

### kubernetes svc

`kubectl get svc kubernetes`

```
NAME         TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
kubernetes   ClusterIP   10.2.0.1     <none>        443/TCP   13d
```

### nmi logs

```
E0102 19:32:10.284772       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.Pod: Get https://10.2.0.1:443/api/v1/pods?fieldSelector=spec.nodeName%3Daks-agentpool-84518513-vmss000000&limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:32:10.285493       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzureAssignedIdentity: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azureassignedidentities?limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:32:10.286265       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzurePodIdentityException: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azurepodidentityexceptions?limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:32:21.287400       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzureAssignedIdentity: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azureassignedidentities?limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:32:21.287678       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.Pod: Get https://10.2.0.1:443/api/v1/pods?fieldSelector=spec.nodeName%3Daks-agentpool-84518513-vmss000000&limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:32:21.288921       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzurePodIdentityException: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azurepodidentityexceptions?limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:32:32.289224       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzureAssignedIdentity: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azureassignedidentities?limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:32:32.291420       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.Pod: Get https://10.2.0.1:443/api/v1/pods?fieldSelector=spec.nodeName%3Daks-agentpool-84518513-vmss000000&limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:32:32.292354       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzurePodIdentityException: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azurepodidentityexceptions?limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:32:43.291544       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzureAssignedIdentity: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azureassignedidentities?limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:32:43.293173       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.Pod: Get https://10.2.0.1:443/api/v1/pods?fieldSelector=spec.nodeName%3Daks-agentpool-84518513-vmss000000&limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:32:43.294368       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzurePodIdentityException: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azurepodidentityexceptions?limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:32:54.293602       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzureAssignedIdentity: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azureassignedidentities?limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:32:54.295845       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.Pod: Get https://10.2.0.1:443/api/v1/pods?fieldSelector=spec.nodeName%3Daks-agentpool-84518513-vmss000000&limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:32:54.297587       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzurePodIdentityException: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azurepodidentityexceptions?limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:33:05.296130       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzureAssignedIdentity: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azureassignedidentities?limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:33:05.299047       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.Pod: Get https://10.2.0.1:443/api/v1/pods?fieldSelector=spec.nodeName%3Daks-agentpool-84518513-vmss000000&limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:33:05.300861       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzurePodIdentityException: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azurepodidentityexceptions?limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:33:16.298310       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzureAssignedIdentity: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azureassignedidentities?limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:33:16.300803       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.Pod: Get https://10.2.0.1:443/api/v1/pods?fieldSelector=spec.nodeName%3Daks-agentpool-84518513-vmss000000&limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:33:16.302125       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzurePodIdentityException: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azurepodidentityexceptions?limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:33:27.300256       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzureAssignedIdentity: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azureassignedidentities?limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:33:27.304225       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.Pod: Get https://10.2.0.1:443/api/v1/pods?fieldSelector=spec.nodeName%3Daks-agentpool-84518513-vmss000000&limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:33:27.305508       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzurePodIdentityException: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azurepodidentityexceptions?limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:33:32.576552       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzurePodIdentityException: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azurepodidentityexceptions?limit=500&resourceVersion=0: EOF
E0102 19:33:32.576630       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.Pod: Get https://10.2.0.1:443/api/v1/pods?fieldSelector=spec.nodeName%3Daks-agentpool-84518513-vmss000000&limit=500&resourceVersion=0: EOF
E0102 19:33:32.576695       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzureAssignedIdentity: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azureassignedidentities?limit=500&resourceVersion=0: EOF
E0102 19:33:43.578621       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzurePodIdentityException: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azurepodidentityexceptions?limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:33:43.579527       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.Pod: Get https://10.2.0.1:443/api/v1/pods?fieldSelector=spec.nodeName%3Daks-agentpool-84518513-vmss000000&limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:33:43.580454       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzureAssignedIdentity: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azureassignedidentities?limit=500&resourceVersion=0: net/http: TLS handshake timeout
E0102 19:33:52.364517       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzureAssignedIdentity: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azureassignedidentities?limit=500&resourceVersion=0: EOF
E0102 19:33:52.364594       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.Pod: Get https://10.2.0.1:443/api/v1/pods?fieldSelector=spec.nodeName%3Daks-agentpool-84518513-vmss000000&limit=500&resourceVersion=0: EOF
E0102 19:33:52.364623       1 reflector.go:205] pkg/mod/k8s.io/client-go@v0.0.0-20180424211516-33f2870a2b83/tools/cache/reflector.go:99: Failed to list *v1.AzurePodIdentityException: Get https://10.2.0.1:443/apis/aadpodidentity.k8s.io/v1/azurepodidentityexceptions?limit=500&resourceVersion=0: EOF
```

## Workarounds
Although it isn't valid in our configuration scenario (production), please let us share a temporary workaround: 

1. allow outbound traffic from the aks subnet cidr to the K8s API Server Public Ip address on 443. 
> Note: the problem with this workaround is that as it is confirmed [here](https://github.com/Azure/AKS/issues/1049#issuecomment-506586554) these IP(s) are not persistent. 
2. another possible workaround could be to start using https://docs.microsoft.com/en-us/azure/aks/private-clusters but it is currently in preview

**Steps To Reproduce**

1. create aks cluster with azure advanced networking 
2. restrict traffic following instructions [here](https://docs.microsoft.com/en-us/azure/aks/limit-egress-traffic#required-ports-and-addresses-for-aks-clusters) + `rt.services.visualstudio.com:443` allowing as source your aks cluster subnet cidr only 
3. install add-pod-identity

**Expected behavior**

no errors should be displayed from logs

**AAD Pod Identity version**

mcr.microsoft.com/k8s/aad-pod-identity/nmi:1.5.4
mcr.microsoft.com/k8s/aad-pod-identity/mic:1.5.4

**Kubernetes version**

client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.0", GitCommit:"e8462b5b5dc2584fdcd18e6bcfe9f1e4d970a529", GitTreeState:"clean", BuildDate:"2019-06-19T16:40:16Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.5", GitCommit:"2640ac46d96791a135961425127d9c2f7e184924", GitTreeState:"clean", BuildDate:"2019-11-14T04:58:54Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}

**Additional context**

EDIT: found this upstream in case it helps https://github.com/kubernetes/client-go/issues/173

EDIT1: just a bit more context on the Azure Firewall SNI `Deny` log entry, since I'm not adding the ASK Api Server Public Ip from my Network Rules (no rules match found), it proceeds to Application Rules, and under that set collection of rules (FQDN), SNI extension could be required.

### PR DESCRIPTION
**Pull request by ferantivero**
solves: #467

**Reason for Change**:

Prescribe add-pod-indentity installation in `kube-system`, otherwise it might fail trying to contact the AKS Api Server

**Issue Fixed**:

fixes #467

**Notes for Reviewers**:

### CODE REVIEW

**Comment by ferantivero** [comment]
I could test this some time next week using helm3.

**Comment by aramase** [comment]
Thank you @ferantivero!

**Comment by aramase** [comment]
@ferantivero Can you please update the PR to add a note instead of changing the deployment to `kube-system`? We can then merge this and close out the issue.

**Comment by ferantivero** [comment]
@aramase sure I can do that. But I‘d rather make the helm chart installation in kube-system the default one.

In my opinion we can get the better of both worlds: production-ready installation by default and backwards compat. 

I‘m initially just adding a note as requested, unless you consider otherwise.

**Comment by aramase** [comment]
@ferantivero That makes sense! So I think we can update this PR to add a note in the deployment section recommending deployment in kube-system namespace for AKS clusters.

Then in a follow up PR, we can add the helm section in here as well https://github.com/Azure/aad-pod-identity#1-create-the-deployment. WDYT?

**Comment by aramase** [comment]
@ferantivero Can you please update the PR when you get the chance? I want to add the recommendation to the docs before more users run into the issue. I can do it too if you aren't able to pick this up? Let me know.

**Comment by ferantivero** [comment]
@aramase done: recommendation has been added to the docs. 

Please let me know in case you want to change the wording, typos, etc.

File: README.md [hunk 1]
```diff
@@ -41,6 +41,8 @@ Or run this command to deploy to a non-RBAC cluster:
 kubectl apply -f https://raw.githubusercontent.com/Azure/aad-pod-identity/master/deploy/infra/deployment.yaml
 ```
 
+> Important: if you are behind a firewall (e.g. Azure Firewall) and the Kubernetes Api Server is outside of your [AKS] cluster subnet, it is recommended to install AAD Pod Identity in the `kuebe-system` namespace by using the [helm charts]. Otherwise, please ensure the Kuberentes Api Server is explicity allowed adding a layer 4 rule to your Firewall.
```
Comment (aramase) [inline README.md:44]: **Important**: For AKS clusters with limited [egress-traffic](https://docs.microsoft.com/en-us/azure/aks/limit-egress-traffic), Please install pod-identity in `kube-system` namespace using the [helm charts](https://github.com/Azure/aad-pod-identity/tree/master/charts/aad-pod-identity). 

@ferantivero wdyt?

Comment (ferantivero) [reply README.md]: done

**Pull request merged**